src/services/services.ts

```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+                    }
+                }
+
+                return includeExtensions ? result : deduplicate(result, (a, b) => a.name === b.name);
+            }
+
+            /**
+             * Check all of the declared modules and those in node modules. Possible sources of modules:
+             *      Modules that are found by the type checker
+             *      Modules from node_modules (i.e. those listed in package.json)
+             *          This includes all files that are found in node_modules/moduleName/ with acceptable file extensions
+             */
+            function getCompletionEntriesForNonRelativeModules(fragment: string, scriptPath: string): CompletionEntry[] {
+                return ts.map(enumeratePotentialNonRelativeModules(fragment, scriptPath), (moduleName) => {
+                    return {
+                        name: moduleName,
+                        kind: ScriptElementKind.unknown,
+                        kindModifiers: ScriptElementKindModifier.none,
+                        sortText: moduleName
+                    };
+                });
+            }
+
+            function enumeratePotentialNonRelativeModules(fragment: string, scriptPath: string) {
+                const ambientSymbolNameRegex = /^"(.+?)"$/;
+
+                // If this is a nested module, get the module name
+                const firstSeparator = fragment.indexOf(directorySeparator);
+                const moduleNameFragment = firstSeparator !== -1 ? fragment.substr(0, firstSeparator) : fragment;
+                const isNestedModule = fragment !== moduleNameFragment;
+
+                // Get modules that the type checker picked up
+                const ambientModules = ts.map(program.getTypeChecker().getAmbientModules(), (sym) => {
+                    const match = ambientSymbolNameRegex.exec(sym.name);
+                    if (match) {
+                        return match[1];
+                    }
+                    // This should never happen
+                    return sym.name;
+                });
+                let nonRelativeModules = ts.filter(ambientModules, (moduleName) => startsWith(moduleName, fragment));
+
+                // Nested modules of the form "module-name/sub" need to be adjusted to only return the string
+                // after the last '/' that appears in the fragment because editors insert the completion
+                // only after that character
+                nonRelativeModules = ts.map(nonRelativeModules, (moduleName) => {
+                    if (moduleName.indexOf(directorySeparator) !== -1) {
+                        if (isNestedModule) {
+                            return moduleName.substr(fragment.lastIndexOf(directorySeparator) + 1);
+                        }
+                    }
+                    return moduleName;
+                });
+
+                forEach(enumerateNodeModulesVisibleToScript(host, scriptPath, moduleNameFragment), visibleModule => {

```

Comment:
Do we plan on supporting completions for classic module resolution as well? I don't currently handle that in the code and that would likely be a pretty expensive check.